scipy/sparse/_coo.py

```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
-            row = np.zeros_like(col)
-        else:
-            raise NotImplementedError(
-                f"coo_matvec not implemented for ndim={self.ndim}")
+        if other.shape != self.shape:
+            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+        other = self.__class__(other)
+        new_data = np.concatenate((self.data, other.data))
+        new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
+        A = self.__class__((new_data, new_coords), shape=self.shape)         
+        return A
+
+    
+    def _sub_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._sub_sparse(self, other)
+
+        if other.shape != self.shape:
+            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+        other = self.__class__(other)
+        new_data = np.concatenate((self.data, -other.data))
+        new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
+        A = coo_array((new_data, new_coords), shape=self.shape)
+        return A
+    
+
+    def _matmul_vector(self, other):
+        if self.ndim<3:
+            result_shape = self.shape[0] if self.ndim > 1 else 1
+            result = np.zeros(result_shape,
+                            dtype=upcast_char(self.dtype.char, other.dtype.char))
+            if self.ndim == 2:
+                col = self.col
+                row = self.row
+            elif self.ndim == 1:
+                col = self.coords[0]
+                row = np.zeros_like(col)
+            else:
+                raise NotImplementedError(
+                    f"coo_matvec not implemented for ndim={self.ndim}")
+
+            coo_matvec(self.nnz, row, col, self.data, other, result)
+            # Array semantics return a scalar here, not a single-element array.
+            if isinstance(self, sparray) and result_shape == 1:
+                return result[0]
+            return result
+        
+        else: # if dim >= 3
+            result = np.zeros(np.prod(self.shape[:-1]),
+                            dtype=upcast_char(self.dtype.char, other.dtype.char))
+            shape = np.array(self.shape)
+            coord = np.concatenate(self.coords)
+            coo_matvec_nd(self.nnz, len(self.shape), shape, coord, self.data,
+                          other, result)
+            
+            result = result.reshape(self.shape[:-1])
+            return result
+
+
+    def _matmul_dispatch(self, other):
+        if self.ndim < 3:
+            return _data_matrix._matmul_dispatch(self, other)
+        
+        M, N = self.shape[-2], self.shape[-1]
+        if other.shape == (N,):
+            return self._matmul_vector(other)
+        if other.shape == (N, 1):
+            result = self._matmul_vector(other.ravel())
+            return result.reshape(M, 1)
+        
+        raise NotImplementedError
+        ("n-D matrix-matrix multiplication not implemented for ndim>2")
```

Comment:
I would keep the `(` on the first line if you are to write it like this, like you have done on line 628. Consider also:
```suggestion
        msg = "n-D matrix-matrix multiplication not implemented for ndim>2"
        raise NotImplementedError(msg)
```

You could consider using this pattern elsewhere to reduce the amount of multi-line strings (but not a big deal either way).